Add effects rework

TeoTwawki opened the pull request:
<!-- place 'x' mark between square [] brackets to affirm: -->
**_I affirm:_**
- [x] that I agree to Project Topaz's [Limited Contributor License Agreement](http://project-topaz.com/blob/release/CONTRIBUTOR_AGREEMENT.md), as written on this date
- [x] that I've _tested my code_ since the last commit in the PR, and will test after any later commits

Still a ton to do before this is ready: see issue #792


Future plans but not in the scope of this pull request: spike effects on armor too. Some ground work is set here but I have made no attempt to transition those at this time.

TeoTwawki committed:
This commit is partly ugly PSEUDO-CODE. This is work in progress and will change and have less redundant bs lines (in theory) when finished..

TeoTwawki committed:
Die in a fire, excess item scripts!

TeoTwawki committed:
part one of many of my being less of an idiot here

TeoTwawki committed:
Recover from bad find-replace screwups

TeoTwawki committed:
Missed one

KnowOne134 committed:
add flame blade Add effects

TeoTwawki committed:
Fix additionalEffectAttack() execution .. sort of.. its running the script again but back to crashing. One or more of the mods is tripping on a nil.

TeoTwawki committed:
Move Modifier ID that was taken in an edit conflict..AGAIN

TeoTwawki committed:
addStatus should have been params.addStatus

TeoTwawki committed:
missed a few that needed the params table prefixed

TeoTwawki committed:
Rename table to avoid confusion and collisions, fix all the mistakes that caused me to notice after doing so.
This mess finally works again and now I can focus moving the redundant parts into smaller functions and smashing the if/else out of existence.

Right now it honestly looks like garbage. Once it doesn't I'll add the item_mods for the weapons and then it can finally be sent to a pull request.

TeoTwawki committed:
Update addEffect functions to reflect the changes of ecbdf23265a04c93e8854f09cd9e163b7ac87861

helixhamin committed:
Remove zanmato_+1.lua and move dispel add effect to item_mods.sql

TeoTwawki committed:
Fix some incorrect comments form a past revision that got missed.

TeoTwawki committed:
Fixed my having accidentally swapped the mnd and int checks.

Also I think at some point I thought I was going to do a params table similar to weaponskills and spells use, but then realized that conflicted with the existing params table being used in the magic resists so I renamed it, but then i scrapped the idea and left "ae" barfed all over. Got rid of that and tabled the functions instead.

Still pending: less if/else insanity...

TeoTwawki committed:
partially revert f3b5b7417cb748123ddb9077502f7af3e8a301ad because I'm retarded now I guess

TeoTwawki committed:
Fix Mjollnir add effect

TeoTwawki committed:
fix typo in comments

TeoTwawki committed:
Cleanup after rebasing onto Topaz.

TeoTwawki committed:
Begin breaking down bloated function..
plan:
 - I am going to need help filling in item_mods.sql before we can call this done.
 - modifiers to control params, not static values in script
 - modular, call only functions we use instead of using if blocks
 - pave the way for eventually handling item spikes here as well (note 1)
 - pave the way for eventually migrating monster additional effect and spikes here as well (note 2)
 - need verification/correction of item's base power (note 3)
 - I have more lines added than removed here and this bothers me.

Note 1:
 - item spikes are currently hard coded into the core

Note 2:
 - attacker/defender, NOT player/mob
 - structure of my current work in progress needs to change a bit so that mobs will be an easy find replace

Note 3:
 - due to issues with the resistance functions, damage amounts and status effect rates will not look like they are retail accurate (low variance).

TeoTwawki commented:
I am now -reasonably- sure I won't be changing the procType enum I am using, if anyone wants to throw item mods to the additional Effect branch I am working off I'd be very grateful. This could go to a tpz branch to be easier to pr too, while I continue to clean up the global script (its functional right now, if mods are added for items) if that is desired. I still have a lot of work to do on the script end to make it maintainable.
```lua
    local procType =
    {
        -- These are arbitrary, make up new ones as needed.
        NORMAL = 1,
        HP_HEAL = 2,
        MP_HEAL = 3,
        HP_DRAIN = 4,
        MP_DRAIN = 5,
        TP_DRAIN = 6,
        HPMPTP_DRAIN = 7,
        DISPEL = 8,
        SELF_BUFF = 9,
        DEATH = 10,
    }
```
```lua
ITEM_ADDEFFECT_TYPE     = 431, -- see procType table in scripts\globals\assitional_effects.lua
ITEM_SUBEFFECT          = 499, -- Animation ID of Spikes and Additional Effects
ITEM_ADDEFFECT_DMG      = 500, -- Damage of an items Additional Effect or Spikes
ITEM_ADDEFFECT_CHANCE   = 501, -- Chance of an items Additional Effect or Spikes
ITEM_ADDEFFECT_ELEMENT  = 950, -- Element of the Additional Effect or Spikes, for resist purposes
ITEM_ADDEFFECT_STATUS   = 951, -- Status Effect ID to try to apply via Additional Effect or Spikes
ITEM_ADDEFFECT_POWER    = 952, -- Base Power for effect in MOD_ITEM_ADDEFFECT_STATUS
ITEM_ADDEFFECT_DURATION = 953, -- Base Duration for effect in MOD_ITEM_ADDEFFECT_STATUS
```
```lua
tpz.subEffect =
{
    -- ATTACKS
    FIRE_DAMAGE         = 1,   -- 110000        3
    ICE_DAMAGE          = 2,   -- 1-01000       5
    WIND_DAMAGE         = 3,   -- 111000        7
    CHOKE               = 3,   -- Shares subeffect
    EARTH_DAMAGE        = 4,   -- 1-00100       9
    LIGHTNING_DAMAGE    = 5,   -- 110100       11
    WATER_DAMAGE        = 6,   -- 1-01100      13
    LIGHT_DAMAGE        = 7,   -- 111100       15
    DARKNESS_DAMAGE     = 8,   -- 1-00010      17
    DISPEL              = 8,   -- Verified with video of Lockheart Greatsword proc.
    SLEEP               = 9,   -- 110010       19
    POISON              = 10,  -- 1-01010      21
    PARALYSIS           = 11,
    AMNESIA             = 11,  -- Verified uses same animation as para
    BLIND               = 12,  -- 1-00110      25
    SILENCE             = 13,
    PETRIFY             = 14,
    PLAGUE              = 15,
    STUN                = 16,
    CURSE               = 17,
    DEFENSE_DOWN        = 18,  -- 1-01001      37
    EVASION_DOWN        = 18,  -- Same subeffect as DEFENSE_DOWN
    ATTACK_DOWN         = 18,  -- Same subeffect as DEFENSE_DOWN
    DEATH               = 19,
    SHIELD              = 20,
    HP_DRAIN            = 21,  -- 1-10101      43
    MP_DRAIN            = 22,  -- This is correct animation
    TP_DRAIN            = 22,  -- Verified this should look exactly like Aspir Samba.
    HASTE               = 23,
    -- There are no additional attack effect animations beyond 23. Some effects share subeffect/animations.
```

TeoTwawki committed:
Figured out what was giving me that sense of unease in my last commit:
repeats of math random and then message set when we fail to proc. Also NIL-ing messages instead of returning out.

Fixes/tweaks/improvements:
- Function now simply returns out when we fail the roll and spares us tons of checking.
- Use elseif instead of if...end (for now, until I migrate the rest of this into self contained modular tabled functions)
- Fixed mods on Zanmato+1 (Because the enum changed since it was added and it was missing its subeffect animation)
- Fixed addEffect Type of Sangarius/Sangarius+1 to be HP drain and not generic damage.
- Removed hard-coded "always proc" used for testing (accidentally slipped into an earlier commit)
- Comment out forced variance for now. Scripts were using math random in addition to resistance to adjust the damage, which shouldn't be done, but the resistance functionality is sort of broken causing your numbers to be pretty consistent.

TeoTwawki commented:
Finally got my latest changes tested.

helixhamin commented:
@TeoTwawki I am adding some "item_mods" 431 additional effect, and since I am not submitting the scripts for these items, the (id,431,1) is just orphaned. I will just add them in using this style, but I need to know how to do it.
Based on your above, for example, item_mods for example, poison.
`INSERT INTO `item_mods` VALUES (16404,431,1);   -- Venom Katars - Additional effect: Poison`

How would I make this a full entry in your style?

TeoTwawki commented:
@helixhamin The new style takes more than one mod. The mods already exist in both release and canary but aren’t used until my pr merges.

In addition to 431, you need 499 (subeffect, this is the animation. Poison animation is 10) 501 (proc rate, 1-100 %) for any and all add effects.

For status such as en-posion, you also need:
951: status effect id
952: power - how big a poison are you inflicting?
953: duration - how long does the poison last?

Those 3 are not needed for damage effects, only status. Damage will use 950 however to tell it which elemental resistance to check, and 500 to tell it how hard to hit.

helixhamin commented:
Okay, so I have:
`function onAdditionalEffect(player,target,damage)
    local chance = 10

    if (math.random(0,99) >= chance or target:isUndead()) then
        return 0,0,0
    else
        target:delStatusEffect(tpz.effect.POISON)
        target:addStatusEffect(tpz.effect.POISON, 4, 3, 30)
        return tpz.subEffect.POISON, tpz.msg.basic.ADD_EFFECT_STATUS, tpz.effect.POISON
    end
end`

This should be converted as follows:
`INSERT INTO `item_mods` VALUES (16404,431,1);   -- Venom Katars - Additional effect: Poison

INSERT INTO `item_mods` VALUES (16404,499,10);  -- Venom Katars - Additional effect: Poison

INSERT INTO `item_mods` VALUES (16404,951,3);   -- Venom Katars - Additional effect: Poison

INSERT INTO `item_mods` VALUES (16404,952,4);   -- Venom Katars - Additional effect: Poison

INSERT INTO `item_mods` VALUES (16404,953,30);  -- Venom Katars - Additional effect: Poison`

Just to make sure this is being done right. Is this correct?

TeoTwawki commented:
You missed 500 for the proc chance. you want:
```sql
-- Venom Katars - Additional effect: Poison
INSERT INTO `item_mods` VALUES (16404,431,1);  -- Additional effect type
INSERT INTO `item_mods` VALUES (16404,499,10); -- Additional effect animation (subeffect)
INSERT INTO `item_mods` VALUES (16404,501,10); -- Additional effect proc chance
INSERT INTO `item_mods` VALUES (16404,951,3);  -- Additional effect status ID
INSERT INTO `item_mods` VALUES (16404,952,4);  -- Additional effect status potency
INSERT INTO `item_mods` VALUES (16404,953,30); -- Additional effect status duration
```

TeoTwawki commented:
You missed 501 for the proc chance. you want:
```sql
-- Venom Katars - Additional effect: Poison
INSERT INTO `item_mods` VALUES (16404,431,1);  -- Additional effect type
INSERT INTO `item_mods` VALUES (16404,499,10); -- Additional effect animation (subeffect)
INSERT INTO `item_mods` VALUES (16404,501,10); -- Additional effect proc chance
INSERT INTO `item_mods` VALUES (16404,951,3);  -- Additional effect status ID
INSERT INTO `item_mods` VALUES (16404,952,4);  -- Additional effect status potency
INSERT INTO `item_mods` VALUES (16404,953,30); -- Additional effect status duration
```

TeoTwawki committed:
Merge branch 'release' of https://github.com/project-topaz/topaz into addEffectsRework

TeoTwawki committed:
Convert Shiva's Claws from old script system to new mod system - kept same values.

Anyone wishing to challenge those values, please make a video+parse your battle log. Several free parser exist for FFXI.

TeoTwawki commented:
The target branch needs to be even with release now, due to shiva's claws.

TeoTwawki commented:
Future items can be added this way:
https://github.com/project-topaz/topaz/pull/572/commits/840993d3f276f99c02cd207bf55b9424be272a7a

TeoTwawki commented:
gdi github

ibm2431 commented:
I'm sorry! 😭

TeoTwawki commented:
@ibm2431 rest api thing fixed it?

ibm2431 commented:
It did, yeah!

ibm2431 commented on scripts/globals/additional_effects.lua:
```diff
@@ -0,0 +1,263 @@
+
+tpz.addEffect.isRanged = function(item)
+    -- Archery/Marksmanship/Throwing
+    return math.abs(item:getSkillType() - tpz.skill.MARKSMANSHIP) < 2
```
That's a neat little trick.

ibm2431 commented on scripts/globals/additional_effects.lua:
```diff
@@ -0,0 +1,263 @@
+    }
+    local delEffect = effectList[addStatus]
+    defender:delStatusEffect(delEffect.strip)
+    return tick
```
tick is null~! (Defender is too, coming into the function)

```lua
local effect = effectList[addStatus]
defender:delStatusEffect(effect.strip)
return effect.tick
```

ibm2431 commented on scripts/globals/additional_effects.lua:
```diff
@@ -0,0 +1,263 @@
+        if element then
+            damage = tpz.addEffect.calcRangeBonus(attacker, defender, element, damage)
+        end
+        chance = ltpz.addEffect.evelCorrection(defender:getMainLvl(), attacker:getMainLvl(), chance)
```
`ltpz`, `evel`

TeoTwawki replied:
wonder how these slipped by travis earlier. the lua linter should have caught it

ibm2431 commented on scripts/globals/additional_effects.lua:
```diff
@@ -0,0 +1,263 @@
+    local element = item:getMod(tpz.mod.ITEM_ADDEFFECT_ELEMENT)
+    local addStatus = item:getMod(tpz.mod.ITEM_ADDEFFECT_STATUS)
+    local power = item:getMod(tpz.mod.ITEM_ADDEFFECT_POWER)
+    local duration = item:getMod(tpz.mod.ITEM_ADDEFFECT_DURATION)
```
Is duration of a status effect always the same when an AE procs? There's no random element?

TeoTwawki replied:
some may get modified by resistance. this is unknown. The current scripts in release are using static duration's with zero modification to them.

ibm2431 commented on scripts/globals/additional_effects.lua:
```diff
@@ -0,0 +1,263 @@
+
+        msgID = tpz.msg.basic.ADD_EFFECT_HP_HEAL
+        attacker:addHP(HP)
+        -- We have to fake this or it will say the defender was healed rather than the attacker.
+        attacker:messageBasic(tpz.msg.basic.ADD_EFFECT_HP_HEAL)
```
If `actor` and `target` get switched by `ADD_EFFECT_HP_HEAL` here, won't they also get switched by `ADD_EFFECT_HP_DRAIN` used by drain types?

(Side: We've got an underlying problem wrong if we need to send out messageBasic to workaround action packet implementation)

TeoTwawki replied:
no because the parameters are correct for those messages. And yes we have an underlying problem. An annoying one thats outside the scope of what I am trying to do to get this globalized. This work around is temporary until the core end allows for some control over the parameters (this is a problem with nearly all messages even outside of additional effect - the core **assumes** what the parameters are going to be and edge cases have to be coded around).

ibm2431 commented on scripts/globals/additional_effects.lua:
```diff
@@ -0,0 +1,263 @@
+            if attacker:hasStatusEffect(tpz.effect.BLINK) or attacker:hasStatusEffect(tpz.effect.UTSUSEMI) then
+                return 0, 0, 0
+            else
+                attacker:addStatusEffect(tpz.effect.BLINK, power, 0, duration)
+                -- We're faking it, so return zeros!
+                msgID = 0
+                msgValue = 0
```
`addStatusEffectEx` in combination with msgID 166 should do the trick.

Contrary to [the comment for ADD_EFFECT_STATUS_3](https://github.com/project-topaz/topaz/blob/c52431e2c48516954aa2f5ebbb331238f43cc3d8/scripts/globals/msg.lua#L240), Windower describes it as:

[166 - Additional effect: \<actor\> gains the effect of \<status\>](https://github.com/Windower/Lua/wiki/Message-IDs)

TeoTwawki replied:
I'll doublecheck this weekend, but i was certain that I was getting told _the mob_ was getting the effect instead of me when I tried all of them.

TeoTwawki replied:
you were right and I've adjusted it

ibm2431 commented on scripts/globals/additional_effects.lua:
```diff
@@ -0,0 +1,263 @@
+    end
+    --------------------------------------
+
+    if addType == procType.NORMAL then
```
In general on this if / elseif block, we could use tabled functions to check for which code to execute in a single operation. Instead of defining local vars from mods at the top, throw them into a table:
```lua
effectInfo = {
    addType = item:getMod(tpz.mod.ITEM_ADDEFFECT_TYPE),
    subEffect = item:getMod(tpz.mod.ITEM_SUBEFFECT),
    damage = item:getMod(tpz.mod.ITEM_ADDEFFECT_DMG),
}
```

Then you can toss them into functions tabled by addType:
```lua
effectActions = {
    procType.NORMAL = function(attacker, defender, effect)
        local msgID, msgValue
        if effect.addStatus and effect.addStatus > 0 then
            local tick = tpz.addEffect.statusAttack(effect.addStatus)
    --etc
        return subEffect, msgID, msgValue
    end
}
```

And then for `additionalEffectAttack` attack itself:
```lua
-- effect is table from above based on item mods
local effectAction = effectActions[item.addType]
subEffect, msgID, msgValue = effectAction(attacker, defender, effect)
return subEffect, msgID, msgValue
```

TeoTwawki replied:
you can see where I started to write in a tabled format for this, but I want to do further refactoring in the branch after the sql is in, but before hitting release. it got too big to handle alone in a single pr. Gonna take several. I promise it will be less "elseif" after.

zach2good replied:
Appreciate the concern in wanting to split your work into multiple saner PRs!

ibm2431 commented on sql/item_mods.sql:
```diff
@@ -20093,6 +20093,11 @@ INSERT INTO `item_mods` VALUES (16550,431,1);
 INSERT INTO `item_mods` VALUES (16557,23,6);
 INSERT INTO `item_mods` VALUES (16557,25,3);
 INSERT INTO `item_mods` VALUES (16562,9,1);
+INSERT INTO `item_mods` VALUES (16564,431,1); -- Flame Blade: AddEffect Type NORMAL
```
On SQL in general, I assume that the item mods for the deleted effect scripts still need to be added?

TeoTwawki replied:
yes. I have an open issue for it, and before the branch is ready a lot of sql work will need done

TeoTwawki committed:
Fix embarrassing last minute mistakes
..Sorry I thought the version I tested was my last committed version..

Xaver-DaRed commented on scripts/globals/additional_effects.lua:
```diff
@@ -118,13 +118,13 @@ function additionalEffectAttack(attacker, defender, baseAttackDamage, item)
             damage = tpz.addEffect.calcRangeBonus(attacker, defender, element, damage)
         end
-        chance = ltpz.addEffect.evelCorrection(defender:getMainLvl(), attacker:getMainLvl(), chance)
+        chance = tpz.addEffect.evelCorrection(defender:getMainLvl(), attacker:getMainLvl(), chance)
```
Missing an "L" evelCorrection

chance = tpz.addEffect.levelCorrection(defender:getMainLvl(), attacker:getMainLvl(), chance)

TeoTwawki replied:
xD

TeoTwawki commented:
I have actually found a few more mistakes on my own, will cleanup this weekend. Hopefully before they get noticed. Wasn't as ready as I thought. it *looked* working in game, but only visually.

TeoTwawki committed:
Verified msgID 166 works like ibm2431 told me,
and adjusted its entry in msg.lua while using it in the additional effect global.
Adjusted deathres checks to be clearer while adding them to additional effect death proc.

TeoTwawki committed:
A few more corrections, plus actually using proper messages.

Unsure why in previous tests the message was saying the monster recovered instead of teh player, but its working correctly now so I'm not going to question it further!

TeoTwawki commented:
ok, except for what I am trying to put to a separate pr after, I think I have this all cleaned up now

(reasoning: I want to enable people to add the sql lines while I kill the if/else blocks without having to wait for me to see that their sql is correct)

59blargedy commented:
@project-topaz/developer  @project-topaz/reviewer  this seems in a good place for a final review, if anyone can assist.

zach2good commented on scripts/globals/additional_effects.lua:
```diff
@@ -0,0 +1,261 @@
+------------------------------------------------------------------------------
+-- This global is intended to handle additional effects from item sources of:
+-- melee attacks, ranged attacks, auto-spikes
+-- Notes:
+-- Ranged versions get bonus from int/mnd, melee does NOT.
+-- No matter how much INT you stack that fire sword doesn't hit any harder.
+-- No matter how much MND you stack that holy mace doesn't hit any harder.
+-- But Ice Arrows and Bloody Bolts will gain damage from INT and Holy bolts will gain damage from MND.
+-- Melee weapon proc also do not appear to adjust for level, only resistance.
+-- In testing my fire sword had the same damage ranges no matter my level vs same mob.
+-- Weakness/resistance to element would swing damage range a lot
+-- For status effects is it possible to land on highly resistant mobs because of flooring.
+------------------------------------------------------------------------------
+require("scripts/globals/teleports") -- For warp weapon proc.
+require("scripts/globals/status")
+require("scripts/globals/magic") -- For resist functions
+require("scripts/globals/utils") -- For clamping function
+require("scripts/globals/msg")
+--------------------------------------
```
Very nicely written up, thanks!

zach2good commented on scripts/globals/additional_effects.lua:
```diff
@@ -0,0 +1,261 @@
+    damage = adjustForTarget(defender, damage, element)
+    damage = finalMagicNonSpellAdjustments(attacker, defender, element, damage)
+
+    --[[
+    This should rightly be modified by resistance checks, and while those DO they are presently not perfect.
+    If you want to force some randomness, un-comment the line below to artificially force 20% variance.
+    ]]
+    -- damage = damage * (math.random(90, 110)/100) --
```
This is helpful!

zach2good commented on scripts/globals/mobskills/perdition.lua:
```diff
@@ -12,7 +12,12 @@ function onMobSkillCheck(target,mob,skill)
 
 function onMobWeaponSkill(target, mob, skill)
-    if target:hasStatusEffect(tpz.effect.MAGIC_SHIELD) or math.random(0,99) < target:getMod(tpz.mod.DEATHRES) then
+    if
+        target:isUndead() or
+        target:hasStatusEffect(tpz.effect.MAGIC_SHIELD) or
+        -- Todo: DeathRes has no place in the resistance functions so far..
+        target:getMod(tpz.mod.DEATHRES) > math.random(100)
+    then
```
Is this meant to be a part of this PR? I don't understand whats going on here

TeoTwawki replied:
While I was adding deathres to the addeffect check, I wanted it to be consistent looking in the existing checks. they work exactly the same just written what I felt was a cleaner/easier to read. And if/when resistance get refactored it can be a single find+replace for all of them this way.

zach2good reviewed (approved):
Finally got around to having a look, I guess I was scared of the giant `-4000 lines` lurking here, looks good 👍

zach2good commented:
Also, thanks for opening up a worklist for what you've planned to do after!
